src/support/istring.h

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2022 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Interned String type, 100% interned on creation. Comparisons are always just
+// a pointer comparison
+
+#ifndef wasm_support_istring_h
+#define wasm_support_istring_h
+
+#include <set>
+#include <string_view>
+#include <unordered_set>
+
+#include <assert.h>
+
+#include "threads.h"
+#include "utilities.h"
+
+namespace wasm {
+
+struct IString {
+private:
+  static std::string_view interned(std::string_view s, bool reuse = true);
+
+public:
+  const std::string_view str;
+
+  IString() = default;
+
+  // TODO: This is a wildly unsafe default inherited from the previous
+  // implementation. Change it?
+  IString(std::string_view str, bool reuse = true)
+    : str(interned(str, reuse)) {}
+
+  // But other C strings generally do need to be copied.
+  IString(const char* str) : str(interned(str, false)) {}
+  IString(const std::string& str) : str(interned(str, false)) {}
+
+  IString(const IString& other) = default;
+
+  IString& operator=(const IString& other) {
+    return *(new (this) IString(other));
```

Comment:
How is this different from the `= default` version?

Reply:
The default implementation uses `operator=` to copy the `str` member, but it's `const` so that doesn't compile and the default implementation is ill-formed and cannot be used.